crates/ruff_python_ast/src/nodes.rs

```diff
@@ -1749,7 +1749,7 @@ impl StringLiteralValue {
 
     /// Returns an iterator over all the [`StringLiteral`] parts contained in this value
     /// that allows modification.
-    pub(crate) fn iter_mut(&mut self) -> IterMut<StringLiteral> {
+    pub fn iter_mut(&mut self) -> IterMut<StringLiteral> {
```

Comment:
The fact that this method isn't public seems intentional to me because mutating can result in an inconsistent to_str result.

Reply:
For my particular usage, I actually only want to mutate the ranges. Thoughts on how to do that in a safe way?

Reply:
In fact, if I had Parser::new_starts_at exposed that would suit my needs even better. I have a string, but I know it starts halfway through a file, and what to parse that.

Reply:
And I guess this only applies to the StringLiteral one - since the others don't have a cached to_str, they could be made public?

Reply:
The parser exposes a parse_expression_range method. I rather not expose the parser struct

Reply:
I need to iterate mutably over FString's for other purposes. Would you be OK exposing the iter_mut for FString and BytesLiteral? Or even just FString? (On the basis BytesLiteral could one day get a to_str, but FString never will)

The parse_expression_range gets a string and parses a subset of it. I instead want to parse the full string, but with an implicit offset. Currently I'm faking that by creating a string of 1000+ spaces, then the expression I care about, and using parse_expression_range. That's pretty inefficient.

Reply:
Do you have some code that you can share. I don't understand the use case and I rather not expose iter_mut because it breaks the type's encapsulation 

Have you considered recreating the entire ExprStringLiteral and cloning the parts? 

For the offset. The main use case for the offsets is so that you can slice into the string but that's not what you want. Have you considered offsetting the ranges e.g when printing the errors or what's the use case for the offsets?

Reply:
I am writing a Python type checker. I've already talked to some Astral folks (Carl - who we should chat with again). Two things:

* In the Python typing spec, you can have strings in place of types. E.g. `def f(x: "list[int]"):`. So you now have a string literal `list[int]` and want to turn it into a type, which is just an expression, so want to call `parse_expression`. But if you can't find `list` etc, you want to report exactly where you couldn't find list. You could of course shift all the ranges (but not unless you expose iter_mut) or shift them for a subset of expressions when printing (but that's super error prone - now you need to wrap every expression with its tracking offset). The easiest thing is to just parse with an implicit offset. Currently we do that by creating a really long string of spaces and using parse_expression_range. You can't just parse from the original string, as there might be things like escapes or multiple literals (which you have to pad out with spaces to get the right positions - which isn't too hard to do).

* Part of our type checker goes walking through expressions to substitute them out with other expressions, to do some book keeping and make some expressions point at other places. Concretely we have `fn expr_visit_mut<'a>(x: &'a mut Expr, f: impl FnMut(&'a mut Expr))` which calls the callback on every one-step recursive function. That's a super handy generic primitive to have. But alas, FString (and only FString) doesn't expose enough pieces to get at the mutable iterators inside. Having `iter_mut` on FString would be valuable and allow such a function. We actually have `expr_visit_mut`, `expr_visit`, `stmt_visit_mut`, `stmt_visit`, `stmt_expr_visit_mut`, `stmt_expr_visit` (the last two finding the expressions inside a statement). It would be wonderful if all those were provided on the AST for free. We use them for loads of things.

Reply:
(in all cases we'd be delighted to supply them as a PR, once we figure out the design)

Reply:
For (1), we do use `parse_expression_range`, you might be interested in going through https://github.com/astral-sh/ruff/blob/5118166d21784e7c78e38ea42919ba50bb2a5142/crates/ruff_python_parser/src/typing.rs which parses string type annotations. Ruff still has a bug for complex type annotations (like implicitly concatenated strings) https://github.com/astral-sh/ruff/issues/10586.

For (2), we have the [`Transformer` trait](https://github.com/astral-sh/ruff/blob/5118166d21784e7c78e38ea42919ba50bb2a5142/crates/ruff_python_ast/src/visitor/transformer.rs) which will visit the nodes to allow mutation.

Is this useful?

Reply:
We are also looking into how to solve 1 for our type checker

Reply:
@dhruvmanila - yep, we have the identical problem, but haven't added the "simple literal" special case. We probably should. But then still end up with the same issue. If you had the parse_expression_at, then the special case becomes non-interesting and you can simplify the code.

The Transformer trait is super interesting. It's a shame that the self isn't mut - that makes accumulating in the visitor much harder. It's a shame there is no lifetime constraint, e.g. if you are walking an `&'a Stmt`, you have the guarantee that all the statements you bump into have the same lifetime `'a` (I note SourceOrderVisitor has this). Can't you use this to screw with the literals in a StringLiteral, thus breaking the `to_str` cache? With those adjustments, I think we'd implement our visitors on top of your Transformer trait.

Reply:
Alternatively, if you added `parse_expression_in_string_literal(x: &StringLiteral) -> Expr` then that would be _exactly_ the primitive we both need, we could delete the slightly dubious `parse_expression_range` (is there another good reason to have this?) and you could do very clever things to deal with things like escapes only once.